### PR TITLE
Store high level pod container image to runtime container annotations.

### LINKED ISSUE #51017: container statuses contain wrong/misleading image tag
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind bug

**What happened**:

When multiple tags of an image share the same SHA256 digest the value status.containerStatuses[].image might return the wrong image tag. Unfortunately `docker inspect` doesn't include the tag that was used to create the container either so this might actually be a bug in docker?

**What you expected to happen**:

status.containerStatuses[].image contains the same tag as specified in spec.containers[].image

**How to reproduce it (as minimally and precisely as possible)**:

 - Create an image named `test:staging` and re-tag it as `test:production`.
 - You now have 2 images with the same name and digest but different tags.
 - Start 2 pods: one using `test:staging` and the other `test:production` as image.
 - The container status for both pods will state that the same image tag was used to create the container which is at least misleading. Although it doesn't really make a difference as both image tags share the same digest it is technically wrong.

**Anything else we need to know?**:

This should imho be considered for #46255.

**Environment**:
- Kubernetes version (use `kubectl version`):
`Client Version: version.Info{Major:"1", Minor:"7", GitVersion:"v1.7.3", GitCommit:"2c2fe6e8278a5db2d15a013987b53968c743f2a1", GitTreeState:"clean", BuildDate:"2017-08-03T07:00:21Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"7", GitVersion:"v1.7.3", GitCommit:"2c2fe6e8278a5db2d15a013987b53968c743f2a1", GitTreeState:"clean", BuildDate:"2017-08-03T06:43:48Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/amd64"}`
- Cloud provider or hardware configuration**: self-hosted cluster
- OS (e.g. from /etc/os-release): `Ubuntu 16.04.3 LTS (Xenial Xerus)`
- Kernel (e.g. `uname -a`): `Linux atgrzsl1751 4.10.0-32-generic #36~16.04.1-Ubuntu SMP Wed Aug 9 09:19:02 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux`
- Install tools: custom ansible playbook
- Others: Docker version: `docker.io (1.12.6-0ubuntu1~16.04.1)`

### PR DESCRIPTION
**Pull request by whypro**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:
Current container runtime implementations set the image in container status to the **first** image from inspect result. This can result in image mismatch when multiple image tags with same hash. (See also: https://github.com/kubernetes/kubernetes/issues/51017)

This PR sets/gets high level container image of a pod to/from runtime container annotations to solve this issue.

**Which issue(s) this PR fixes**:

Fixes #51017

**Special notes for your reviewer**:

Thanks for @loveRhythm1990 's previous work (https://github.com/kubernetes/kubernetes/pull/92249) and also thanks for @hex108 's advise.

**Does this PR introduce a user-facing change?**:

```release-note
Ensure image tag correctly in container statuses when two images have same hash.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by whypro** [comment]
cc @hex108  @zhouya0 
PTAL, thanks!

**Review by zhouya0 — APPROVED** [review]

**Comment by zhouya0** [comment]
/lgtm
Thanks for the fix!

**Comment by loveRhythm1990** [comment]
/lgtm
Thanks for the fix!, since I am busy with other things,  I will close my origin PR, hope this problem been fixed soon.

**Comment by hex108** [comment]
/lgtm

Thanks!

**Comment by whypro** [comment]
ping @Random-Liu @yujuhong . 
PTAL, thanks!

**Comment by whypro** [comment]
cc @derekwaynecarr @sjenning

**Comment by sjenning** [comment]
This seems reasonable since there is no way to disambiguate without storing this information from the spec (as far as I can tell).

@mrunalp what do you think?

**Comment by sftim** [comment]
For the issue in https://github.com/kubernetes/kubernetes/issues/51017 I'd expect that the kubelet could disambiguate which container name and tag are used, based on the `.spec` of the Pod that the container belongs to.

https://github.com/kubernetes/kubernetes/issues/51017 implies that I'm wrong though. When can't the kubelet use the `.spec` to work out the right answer?

**Comment by mrunalp** [comment]
Will review this tomorrow.

**Comment by derekwaynecarr** [comment]
can we add a node e2e?

**Review by derekwaynecarr — CHANGES_REQUESTED** [review]
I have no objection to the new label.

Can you add a node e2e test so we have something that verifies this does not regress?

**Comment by whypro** [comment]
@derekwaynecarr OK. I'll add e2e tests later.

**Comment by hex108** [comment]
Hi @whypro, is there any progess here?

**Comment by ehashman** [comment]
/hold

please remove the hold once you've added tests to flag a reviewer's attention, thanks!

**Comment by whypro** [comment]
@derekwaynecarr @hex108 @ehashman 

Sorry for the delay because I am busy with other things.

I have added the e2e test, but the test relies on an image of `nginx:1.14-1-copy`. I know from the [documentation](https://github.com/kubernetes/k8s.io/tree/main/k8s.gcr.io) that I need to upload it to `k8s.gcr.io/k8s-staging-e2e-test-images`, then promote to `k8s.gcr.io/e2e-test-images` repo.

But it seems that I didn't have push permission, would you like to tell me how to do it or help me push it?

**Comment by claudiubelu** [comment]
I think the best approach would to to separate the bugfix from the test, so we can at least proceed with the bugfix and worry about the test later. It seems that the Image Promoter can't really handle 2 image tags having the same SHA.

**Comment by claudiubelu** [comment]
> 
> 
> I think the best approach would to to separate the bugfix from the test, so we can at least proceed with the bugfix and worry about the test later. It seems that the Image Promoter can't really handle 2 image tags having the same SHA.

The ``nginx:1.14-1-copy`` image has been pushed and it has the same SHA as the ``nginx:1.14-1`` image (https://github.com/kubernetes/k8s.io/pull/2351). This only needs to be rebased.

**Comment by whypro** [comment]
@claudiubelu Thanks very much for help! I've rebased the branch and it now is ready for review.
cc @derekwaynecarr @Random-Liu @yujuhong

**Comment by ritpanjw** [comment]
Hi @dchen1107 👋 I'm checking in from the bug triage team for release 1.24. Is this PR targeted for release 1.24?

**Comment by ritpanjw** [comment]
Hi @dchen1107 👋 I'm checking in from the bug triage team for release 1.24. we are very near to code freeze, any updated?

**Comment by dims** [comment]
@derekwaynecarr @mrunalp 

Do we want this in 1.24?

File: pkg/kubelet/kuberuntime/kuberuntime_container.go
Comment (ehashman) [inline pkg/kubelet/kuberuntime/kuberuntime_container.go:564]: This will always override the `status.Image.Image` above if the annotated info is present, is that what we want?

Comment (pacoxu) [reply pkg/kubelet/kuberuntime/kuberuntime_container.go:564]: Before the new label is set, we keep using `status.Image.Image`.
If we can get from the label(info annotation), we will use it instead.
/lgtm

File: test/e2e/node/pods.go
Comment (oomichi) [inline test/e2e/node/pods.go:513]: Does this `NginxCopy` image have the same hash as the original `Nginx` image to reproduce the issue?

**Review by oomichi — COMMENTED** [review]
/cc @oomichi

File: test/e2e/node/pods.go
Comment (ehashman) [inline test/e2e/node/pods.go:524]: I think it's possible for this to race. Won't there be a lag between the update of the container image and the propagation of the status with new image?

**Review by ehashman — COMMENTED** [review]
This was last tested so long ago that there are no results in prow.

/test pull-kubernetes-e2e-gce-ubuntu-containerd

**Comment by ehashman** [comment]
/milestone v1.25

Would like to see this land but I think we have other priorities right now for code freeze. Let's try to land this early in 1.25.

File: test/e2e/node/pods.go
Comment (claudiubelu) [reply test/e2e/node/pods.go:513]: ``NginxCopy`` (``nginx-1.14.2-copy``) has the same SHA as the ``Nginx`` image (``nginx-1.14.2``), according to the SHAs promoted: https://github.com/kubernetes/k8s.io/blob/c011745da40d143dfbec89737901143cfed154e7/k8s.gcr.io/images/k8s-staging-e2e-test-images/images.yaml#L78

File: test/e2e/node/pods.go
Comment (whypro) [reply test/e2e/node/pods.go:524]: Yes, there will be a lag. `WaitForPodCondition` will poll until two image are consistent.

**Comment by whypro** [comment]
Forgive me for taking a long time to response.
@claudiubelu @helayoty @ehashman PTAL and review this PR again, thanks.

**Comment by dims** [comment]
If you still need this PR then please rebase, if not, please close the PR

**Comment by vinaykul** [comment]
> pull-kubernetes-e2e-inplace-pod-resize-containerd-main-v2

Please ignore this test. The failure is because InPlacePodVerticalScaling feature gate is not defined.
This test has now been limited to executing only on PR https://github.com/kubernetes/kubernetes/pull/102884 for the time being.

**Comment by bart0sh** [comment]
@vinaykul 
> > pull-kubernetes-e2e-inplace-pod-resize-containerd-main-v2
> 
> Please ignore this test. The failure is because InPlacePodVerticalScaling feature gate is not defined. This test has now been limited to executing only on PR #102884 for the time being.

I saw this test failing in many unrelated PRs. It confuses contributors. Can we trigger it only for the code related to inplace pod vertical scaling?

**Comment by sftim** [comment]
> **Does this PR introduce a user-facing change?**:
> 
> ```
> NONE
> ```

Is that right? I'd expect that end users would be able to notice that the change had been made.

**Comment by whypro** [comment]
Thanks for notice. I've added it to release-note.

**Comment by vinaykul** [comment]
> @vinaykul
> 
> > > pull-kubernetes-e2e-inplace-pod-resize-containerd-main-v2
> > 
> > 
> > Please ignore this test. The failure is because InPlacePodVerticalScaling feature gate is not defined. This test has now been limited to executing only on PR #102884 for the time being.
> 
> I saw this test failing in many unrelated PRs. It confuses contributors. Can we trigger it only for the code related to inplace pod vertical scaling?

@bart0sh I believe the fix is already in. @bobbypage submitted PR https://github.com/kubernetes/test-infra/pull/28229 that limits the trigger.

Job history shows the change is effective, https://prow.k8s.io/job-history/gs/kubernetes-jenkins/pr-logs/directory/pull-kubernetes-e2e-inplace-pod-resize-containerd-main-v2

A retest command though would trigger it even in unrelated PRs because a failed test history has been captured. Please use retest-required for now or ignore this particular test.

Once we merge the API change (hopefully soon) and the FG definition, this should start succeeding as a no-op.

**Comment by vinaykul** [comment]
/skip pull-kubernetes-e2e-inplace-pod-resize-containerd-main-v2

**Comment by whypro** [comment]
ping @derekwaynecarr . Could you review changes again?

**Comment by SergeyKanzhelev** [comment]
/assign @ruiwen-zhao 

@ruiwen-zhao mostly FYI

File: test/e2e/node/pods.go
Comment (SergeyKanzhelev) [inline test/e2e/node/pods.go:270]: is the test always failing without this fix? does runtime uses already existing image in status?

File: test/utils/image/manifest.go
Comment (SergeyKanzhelev) [inline test/utils/image/manifest.go:191]: can you please extend the comment to make sure there it's clear we want the same SHA here, but different tag

**Comment by SergeyKanzhelev** [comment]
Depending on what will be merged faster, the ContainerStatus docs will needs to be updated. Like this: https://github.com/kubernetes/kubernetes/pull/115463

**Comment by SergeyKanzhelev** [comment]
/lgtm cancel

Cancelling to make sure the comment on nginx copy image is updated before the merge.

**Comment by ruiwen-zhao** [comment]
One thing to call out here is that containerd prefixes some image names with `docker.io/library`.

Currently, without this change, if you deploy a container that runs `nginx:1.14.2`, then you will see  `docker.io/library/nginx:1.14.2` in the container status. But after this change, container status will have `nginx:1.14.2`. 

I don't think this is something we should be concerned about, but just calling out the change of behavior.

**Comment by sftim** [comment]
> One thing to call out here is that containerd prefixes some image names with `docker.io/library`.
> 
> Currently, without this change, if you deploy a container that runs `nginx:1.14.2`, then you will see `docker.io/library/nginx:1.14.2` in the container status. But after this change, container status will have `nginx:1.14.2`.

Good spot. We shouldn't hard code or imply a presumption about how to canonicalize an image name.

**Pull request closed without merging**
